{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3ef63651",
   "metadata": {},
   "source": [
    "# Project: Visualizing Cell Differentiation\n",
    "\n",
    "### 1. Background : Stem Cell Differentiation\n",
    "Every biological cell contains the same \"source code\" (DNA). However, a heart cell behaves differently from a neuron because they \"execute\" different parts of that code. In other words, genes are expressed at different levels. In order for cells to reach their fates ('final' type), they have to undergo a journey, the so called **cell differentiation** or **cell specification**. \n",
    "\n",
    "Biologists often use the **Waddington Landscape** analogy to describe the process of cell differentiation: imagine a ball rolling down a hilly terrain. At the top, the ball can go down many different valleys. By the bottom, the ball has journeyed through a specific path and is stuck in a specific state.\n",
    "\n",
    "![alt text](./images/waddington2.png)\n",
    "\n",
    "\n",
    "### 2. Data\n",
    "\n",
    "We quantify this execution by measuring the levels of **Gene Expression** throughout development.\n",
    "\n",
    "In the current project, the data follows **Mouse Embryonic Stem Cells (ESCs)** over a period of 7 days (168 hours). \n",
    "At $T=0$, the cells are \"Pluripotent\"â€”we can think of them as uninitialized objects that have the potential to become almost any tissue type. As time progresses, they interprete gene expression levels as signals to **differentiate**. At the end of the experiement, we get two cell types: **E14** and **R1**.\n",
    "\n",
    "The **Single-Cell Gene Expression** dataset at our disposal is the following. \n",
    "*   **Dimensions:** Each cell is a data object described by **96 features** (different marker genes, in addition to time and type of cell).\n",
    "\n",
    "| Feature | Type | Description |\n",
    "| :--- | :--- | :--- |\n",
    "| **Time** | `int` | The timestamp of the snapshot (0h, 24h, 48h, 72h, 96h, 120h, 168h). |\n",
    "| **Type** | `string` | The cell line background (**E14** or **R1**). These are two different \"classes\" of stem cells. |\n",
    "| **Genes** | `float` | **96 columns** (e.g., *Actb, Nanog, Sox2...*). These represent the normalized expression level of specific genes. |\n",
    "\n",
    "\n",
    "*** \n",
    "*Data Source & Paper: Patrick S. Stumpf et al., \"Stem Cell Differentiation as a Non-Markov Stochastic Process,\" Cell Systems, 2017.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6ba993a",
   "metadata": {},
   "source": [
    "### Part 1. Load the dataset\n",
    "Read the \"gene_expression_data.csv\" dataset into a pandas dataframe. \n",
    "\n",
    "Once loaded, you can use the \"head()\", \"info()\" or \"describe()\" functions for an overview of the dataset and its summary statistics. Below is an example result of head() applied to the dataset.\n",
    "\n",
    "![alt text](./images/head.png)\n",
    "\n",
    "Tip: Because the dataset, although high-dimensional, only has 96 columns, you can also open the dataset using Excel and view it in tabular form by using the the \"text-to-columns\" feature of MS Excel (https://www.youtube.com/watch?v=QyZ6IMkln2U)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc7265ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Part 1. Write your code in this cell ############"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "416bd9e5",
   "metadata": {},
   "source": [
    "### Part 2. Visualize some data\n",
    "Because the dataset is high-dimensional (cells are described across 96 genes), we canNOT visualize all genes. In this part, we will restrict our focus on two genes: \"Bmp4\" and \"Nanog\".\n",
    "\n",
    "Create a figure with two plots (one for each gene), where each plot shows, at each time (0H, 24H, ..., 168H), the statistics of expresssion levels of the genes in cells. What plot types are appropriate for this ?\n",
    "Hint: you may want to have time on the x-asis  \n",
    "\n",
    "Bonus: There are two types of cells in the dataset (E14 and R1). Can you enhance the plot to visualize the expression levels within each cell type ? What do you observe ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39343672",
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Part 2. Write your code in this cell ############"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e591f06b",
   "metadata": {},
   "source": [
    "### Part 3. Relationships between genes\n",
    "Because the dataset is high-dimensional (cells are described across 96 genes), we canNOT visualize all relationships between our variables (genes). However, as in almost every real-world datasets, variables entertain strong or weak relations. In this part, we will try to identity which genes that have strong ties, and attempt to visualize how the behave together.\n",
    "\n",
    "Once more, because the dataset is multi-dimensional, we will restrict our analysis to the subset of genes (although we can get away with an analysis of the whole set of genes in our current scenario of \"just\" 96 genes) \n",
    "\n",
    "    genes = ['Nanog', 'Pou5f1', 'Sox2', 'Gata6', 'Pax6', 'Sox1', 'Actb', 'Bmp4'].\n",
    "\n",
    "Create a figure with the following plots:\n",
    "\n",
    "1. A **correlation heatmap** showing the correlations of all genes above at time O\n",
    "2. A **correlation heatmap** showing the correlations of all genes above at time 168\n",
    "3. From the first plot, pick the two genes with the highest correlation and create a **scatter plot** of both. Does the scatter trend verify the observed correlation ? Compare their correlation at time 168. What do you observe ?\n",
    "4. From the second plot, pick the two genes with the highest correlation and create a **scatter plot** of both. Does the scatter trend verify the observed correlation ? Compare their correlation at time 0. What do you observe ?\n",
    "\n",
    "Bonus: i.Think about improvements you could make on the scatter plots. What comes to mind ? Create a figure with two plots (Hint: different colors and/or markers for the two different cell types, etc.)\n",
    "ii. How does the relationship of the 2 genes picked at (3.) evolve through all time points (0,24,48,72,69,120,144,168) ? What plot(s) can you use for this ? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4654f3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Part 3. Write your code in this cell ############"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a229dcf7",
   "metadata": {},
   "source": [
    "### Part 4. Dimensionality Reduction (PCA)\n",
    "We can only do so much in trying to visualize our dataset one variable at the time, or 2 variables at the time. Our dataset is high-dimensional, we must therefore use appropriate methods to visualize it. Most notably, dimensionality reduction methods.\n",
    "\n",
    "1. Perform **Principal Component Analysis (PCA)** on the data, keeping only 2 dimensions\n",
    "(Hint: you don't need to pass columns 'Time' and 'Type' to PCA as they do not contain information on genes. Extract just genes columns and pass it to PCA. The line below shows how you can 'drop' those columns)\n",
    "\n",
    "    X = df.drop(['Time', 'Type'], axis=1)\n",
    "\n",
    "2. Determine the 'importance' of each of your principal component ? How do you inteprete that ?\n",
    "\n",
    "3. Create a scatter plot on the resulting two dimensional data (PC1 vs PC2). What do you observe ? Can you identify different trajectories in time for different cell types (E14, R1) ?\n",
    "\n",
    "4. Enhance the scatter plot using different marker types for cell type (E14, R1) and different colors for time (0, 24, ..., 268). What do you observe ?\n",
    "\n",
    "Bonus: Perform **Principal Component Analysis (PCA)** on the data, this time keeping 3 dimensions. Redo steps 1-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68497caf",
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Part 4. Write your code in this cell ############"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3228b75",
   "metadata": {},
   "source": [
    "### Part 5. Dimensionality Reduction (Non-linear methods)\n",
    "Non-linear dimensionality reduction (t-SNE, UMAP) methods usually produce better visualisations than PCA\n",
    "\n",
    "1. Redo steps 1,3,4 of the previous part (keeping only 2 components, then 3 components) using a non-linear reduction method of your choice (**t-SNE** or **UMAP**). What do you observe ? How is the visulisation produced compared to the visual result of PCA ?\n",
    "\n",
    "\n",
    "Bonus: Redo step 1 using the dimensionality reduction method **isomap** (https://scikit-learn.org/stable/modules/generated/sklearn.manifold.Isomap.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43f1333c",
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Part 5. Write your code in this cell ############"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0f60432",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
